v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+        (void)memcpy(pos, "content", sizeof("content")); pos += sizeof("content");
+        *pos = TlgInANSISTRING;  pos++;
+
+        /* Codes_SRS_LOG_SINK_ETW_01_036: [ The string file (as field name, excluding zero terminator), followed by one byte with the value TlgInANSISTRING. ]*/
+        /* Codes_SRS_LOG_SINK_ETW_01_045: [ Length of the file field name (determined at compile time, excluding zero terminator) + 1. ]*/
+        (void)memcpy(pos, "file", sizeof("file")); pos += sizeof("file");
+        *pos = TlgInANSISTRING;  pos++;
+
+        /* Codes_SRS_LOG_SINK_ETW_01_037: [ The string func (as field name, excluding zero terminator), followed by one byte with the value TlgInANSISTRING. ]*/
+        /* Codes_SRS_LOG_SINK_ETW_01_046: [ Length of the func field name (determined at compile time, excluding zero terminator) + 1. ]*/
+        (void)memcpy(pos, "func", sizeof("func")); pos += sizeof("func");
+        *pos = TlgInANSISTRING;  pos++;
+
+        /* Codes_SRS_LOG_SINK_ETW_01_038: [ The string line (as field name, excluding zero terminator), followed by one byte with the value TlgInINT32. ]*/
+        /* Codes_SRS_LOG_SINK_ETW_01_047: [ Length of the line field name (determined at compile time, excluding zero terminator) + 1. ]*/
+        (void)memcpy(pos, "line", sizeof("line")); pos += sizeof("line");
+        *pos = TlgInINT32;  pos++;
+
+        uint16_t metadata_size_without_properties = (uint16_t)(pos - &self_described_event->metadata[0]);
+
+        if (property_value_count <= LOG_MAX_ETW_PROPERTY_VALUE_PAIR_COUNT)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_053: [ For each property in log_context the following shall be added to the event metadata: ]*/
+            for (i = 0; i < property_value_count; i++)
+            {
+                /* Codes_SRS_LOG_SINK_ETW_01_051: [ For each property in log_context, the length of the property name + 1 and one extra byte for the type of the field. ]*/
+                size_t name_length = strlen(context_property_value_pairs[i].name);
+                /* Codes_SRS_LOG_SINK_ETW_01_085: [ If the size of the metadata and the formatted message exceeds 4096 bytes, log_sink_etw.log_sink_log shall not add any properties to the event. ]*/
+                if ((uint16_t)(&self_described_event->metadata[MAX_METADATA_SIZE] - pos) < name_length + 1 + 1)
+                {
+                    add_properties = false;
+                    break;
+                }
+                else
+                {
+                    /* Codes_SRS_LOG_SINK_ETW_01_054: [ A string with the property name (including zero terminator) ]*/
+                    (void)memcpy(pos, context_property_value_pairs[i].name, name_length + 1);
+                    pos += name_length + 1;
+                }
+
+                /* Codes_SRS_LOG_SINK_ETW_01_055: [ A byte with the type of property, as follows: ]*/
+                switch (context_property_value_pairs[i].type->get_type())
```

Comment:
For now I will leave it here, otherwise I'd have to push the tlg enums into the context. Maaaaaybe that will come at some point, but likely this won't be our top issue now.